### DISCUSSION TITLE
인프라 아키텍처 개선하기 Part.2

### DISCUSSION THREAD

**Discussion by seokhongkim** [woowacourse/infra-architecture-4#4] [General]
팀에서 아래 내용을 함께 논의한 후 정리해서 코멘트로 남겨주세요.

데이터와 트래픽이 많아질 때의 상황을 가정하고, 가정한 내용을 포함해서 작성해주세요.
### 예시
1. 서비스 사용자가 1,000만명으로 늘어날 때 상품의 개수는 50만개가 등록되어 있고, 상품당 리뷰는 1만개가 등록된다.
2. 서비스 사용자가 1,000만명으로 늘어난 시점에 글의 개수는 5,000만개가 등록되어 있고, 일 평균 50만건의 글이 새로 등록된다. 피크 타임에 전체 글 중 40%의 글이 등록된다.

## 코멘트 형식

# 팀 이름
- 개선한 인프라 아키텍처 이미지
- 인프라 아키텍처를 변경하면서 고민한 부분
  - 서비스 데이터의 특성 (위의 예시와 같이 가정한 내용)
  - 우리 서비스에서 특별히 고려해야 할 중요한 기능
    - e.g., 문자열 검색이 되어야 한다, 사용자에게 실시간으로 알람을 보내줘야 한다, 사용자가 동시에 데이터를 변경할 수 없도록 막아야 한다, 메인 페이지에서 특정 데이터를 자주 조회한다 등
  - 개선한 인프라 아키텍처에서 발생할 수 있을 것으로 예상되는 문제
  - 문제에 대한 해결 방안 아이디어

**Comment by tco0427** [comment]
### 모아모아 팀 인프라 아키텍처 개선하기

![IMG_0320](https://user-images.githubusercontent.com/57028386/194460024-1e5f27c9-c57b-4892-92b1-4e0fe677522b.jpg)

- 인프라 아키텍처를 변경하면서 고민한 부분
    - 서비스 데이터의 특성
        - 글, 이미지
    - 우리 서비스에서 특별히 고려해야 할 중요한 기능
        - 알림 서비스
            - 비동기로 처리한다. (메지시큐)
                
                실패했을 때는? → 실패 여부를 DB에 저장해두고, 실패에 대해서만 10분 간격으로 배치처리를 한다.
                
        - 메인페이지에 커서 적용
            
            무한 스크롤이기 때문에, 마지막에 조회할 때 사용한 where절의 조건을 재사용한다.
            
        - 10억건의 스터디가 생성된다면??
            - 메인 페이지에서 인기순으로 보여주게 될 것이고, 캐싱을 적용해서 일정 시간마다 업데이트해준다.
            - 인기순이 하트 순이든 조회순이 된다면 역정규화도 이용해볼 수 있을 것 같다.
        - 스터디 가입
            
            여러 WAS로 동일한 스터디에 대한 요청이 왓을 때, 메시지 큐를 통해서 순서를 보장받고, 순차적으로 DB에 업데이트한다. (생성자-소비자 패턴)
            
    - 개선한 인프라 아키텍처에서 발생할 수 있는 것으로 예상되는 문제
        - 비동기로 처리하므로 문제점을 찾기가 어려우므로 모니터링을 할 수 있도록 구성해야한다.

**Comment by ulimy** [comment]
# 회고덕
- 개선한 인프라 아키텍처 이미지
    ![IMAGE](https://user-images.githubusercontent.com/18046394/194460109-6794b323-9487-47fc-8c2a-ef3b0cc5f66d.jpeg)


- 인프라 아키텍처를 변경하면서 고민한 부분
    - cud보다 r 연산이 약 10배 이상 많다.
- 개선 방향
    - 파티셔닝/샤딩이 가능한 요소 
        - 회고 생성일 기준 분리
        - 템플릿 관련 데이터
        - 장점 ) 조회 성능이 향상될 수 있다.
        - 단점 ) 여러개의 파티션에 걸쳐 조회하는 경우 성능이 저하될 수 있다. ( 회고폼을 기반으로 작성된 회고를 모아보는 경우 과거부터 현재까지의 회고가 모여있을 수 있어 해당 문제가 발생할 수 있을 것 같다 )
    - DB 다중화 ( Replication )
    - 캐싱
        - 인기있는 템플릿의 경우 수정, 삭제가 자주 일어나지 않으며 고정되어 있을 확률이 높기 때문에 적절한 요소라고 판단된다.
    - 로드밸런싱

**Comment by hamcheeseburger** [comment]
# F12

- 개선한 인프라 아키텍처 이미지
    
![KakaoTalk_Photo_2022-10-07-12-13-47](https://user-images.githubusercontent.com/61769743/194460413-5b248009-bb68-4136-a0aa-acc4bc0ebf85.jpeg)

    
- 인프라 아키텍처를 변경하면서 고민한 부분
    - 서비스 데이터의 특성
        - 일당 1000만명, 리뷰작성은 10만건, 팔로잉은 10만건
    - 우리 서비스에서 특별히 고려해야 할 중요한 기능
        - 제품과 사용자에 대한 문자열 검색이 가능해야한다.
        - 사용자간 팔로잉이 가능해야 한다.
        - 메인페이지에서 인기상품과 최신 리뷰를 자주 조회한다.
    - 개선한 인프라 아키텍처에서 발생할 수 있을 것으로 예상되는 문제
        - 하루에 10만건씩 리뷰 작성이 된다면, 결국에 DB가 꽉 차는 상황이 올 것이다.
        - 데이터 정합성을 위해 배치 주기를 정하기가 애매하다.
    - 문제에 대한 해결 방안 아이디어
        - 리뷰 데이터가 1개의 DB에 가득 차는 상황이 온다면, 제품의 카테고리로 전체적인 샤딩이 진행되고 제품 id의 range를 기준으로 세부적인 샤딩을 진행할 것이다.

**Comment by YJGwon** [comment]
## 체크메이트 팀 아키텍쳐 개선

![image](https://user-images.githubusercontent.com/67885363/194460496-e43da34b-f670-4008-b71e-b87642528bc8.png)

### 우리 서비스의 특성

- 아침에 출석이 몰린다
    - 우테코가 아니더라도 일반적으로 출석을 많이 하게 될 등교, 출근시간 등
- 출석 체크 요청은 최대한 사용자의 action을 latency 없이 반영해야 한다

---

### scalability

- 1000만명 감당하려면 일단 하드웨어 스펙은 올려야 하지 않을까 (t4g micro는 쫌..)
- Nginx는 scale-out보다는 scale-up으로 개선하는게 좋지 않을까?

### performance

- 지각 횟수 역정규화
    - 지각 횟수는 출석 row를 count해서 도출하는데, 출석 데이터는 시간과 비례하게 계속 늘어날 데이터
- 회원 data 검색 엔진
    - 회원 data는 CUD보다 R이 훨씬 많을 것으로 예측됨
    - 조회 조건과 결과는 굉장히 random하기 때문에 cache는 부적절
    - 검색 엔진을 통해서 CUD 성능을 약간 희생하더라도 R 성능 개선
- 출석 요청 Message Queue
    - 출석 요청 처리를 매번 DB 작업 완료한 후 응답을 주는 것이 아니라, 출석을 요청한 시간과 함께 출석 처리에 대한 Message를 Queue에 담아서 DB 처리는 비동기적으로
    - 출석 요청은 최대한 latency없이 accept되어야 하기 때문에 중요한 개선포인트!
- 모임 참가자에 대한 지각 횟수 Cache
    - 보통 오전에 출석을 하는 경우가 대다수 일것이라고 예상됨
    - 한 번 출석이 마감되면, 남은 시간동안 그 날의 지각 횟수는 변하지 않는다
    - 정오에 batch를 통해 각 모임의 출석 현황, 지각 횟수 등을 조회한 다음 그 결과를 cache
- attendance 또는 event table에 대한 샤딩, 파티셔닝
    - 샤딩을 시간 기준으로 하게 되면 하드웨어가 충분히 활용되지 않을 것 같다
    - 샤딩은 User나 Meeting 기준으로 해서 용량도 충분히 사용하게 하고 트래픽도 적절히 분산되게 하는 게 좋을 것 같다
    - 파티셔닝은 시간을 기준으로 해서 조회 성능을 upgrade할 수 있을 것 같다
    - 샤딩을 어떻게 하면 잘 활용할  수 있을지 아직은 잘 모르겠다 ㅠㅠ

**Comment by Seongwon97** [comment]
# 모두 모여라 아키텍처
![image](https://user-images.githubusercontent.com/57744251/194460767-4cca5bd0-d622-4f03-9801-960bf83e7d62.png)

- 개선한 인프라 아키텍처 이미지
- 인프라 아키텍처를 변경하면서 고민한 부분
    - 서비스 데이터의 특성 (위의 예시와 같이 가정한 내용)
        - 모임에 대해서는 변경에 비해 조회의 개수가 많을 것이라고 가정
        - 트래픽이 많아질 경우에 대해 트랜잭션이 유실 가능성이 존재
        - 모임 검색이 빠르게 되어야 한다.
    - 우리 서비스에서 특별히 고려해야 할 중요한 기능
        - 문자열을 통한 모임 검색이 되어야 한다.
            - 현재 Like를 통해 검색을 하여 ES를 통해 성능 향상을 할 수 있을 것 같다.
        - 동시에 참여요청이 왔을 때, 참여자가 모임의 참여 인원 수를 초과하면 안된다.
        - 메인 페이지의 모임 목록 조회 데이터를 자주 조회한다.
    - 개선한 인프라 아키텍처에서 발생할 수 있을 것으로 예상되는 문제
        - 캐시 불일치 처리에 대해 아직 미비..하다
            - 캐시 불일치를 처리하는 비용 vs DB에 직접 접근하여 조회하는 비용
        - Queue 크기보다 더 많이 요청이 들어온다면 어떻게 할까?

**Comment by DWL21** [comment]
# 꼭꼭
![IMG_3653 (1)](https://user-images.githubusercontent.com/76774809/194460819-2bffbd5a-d166-406b-a255-7be35b955f43.jpg)

- 우리 서비스에서 특별히 고려해야 할 중요한 기능
  - 사용자에게 슬랙 푸시 알림을 보낸다.
  - 닉네임 기반 사용자 검색을 한다.
  - 쿠폰 상태가 변해 정합성을 생각해야 한다.
  - DB를 나누면 쿠폰을 보관하는 기준을 정해야 한다.
- 문제에 대한 해결 방안 아이디어
  - 푸시 알림은 메세지 큐를 사용한다.
  - 시간 주기로 보내는 알림은 배치 프로그램을 이용한다.
- 개선한 인프라 아키텍처에서 발생할 수 있을 것으로 예상되는 문제
  - 배치 단위를 길게 가져가면 알림을 보내지 않는 조건으로 데이터가 변경되는 경우가 있다.

**Comment by hyunrrr** [comment]
# 속닥속닥
![IMG_4046](https://user-images.githubusercontent.com/54316950/194461039-a0ce8acb-c571-4698-a2db-e1111ead345c.png)
- 이미지 서버 앞에 캐시서버를 둠
    - 자주 사용되는 이미지에 빠르게 접근할 수 있게 함
- 외부라이브러리 사이에 messageQueue를 둠
    - 요청이 몰렸을때 messageQueue로 트래픽 관리
    - 회원가입 시 이메일을 전송(smtp)
-  DB에서 자주 읽는 데이터 캐싱
    - 메인페이지 데이터 캐싱
- master DB 중간에 messageQueue를 둠
    - 쓰기 부하가 늘어날때 messageQueue를 통해 완화

**Comment by sure-why-not** [comment]
# 땡쿠
![image](https://user-images.githubusercontent.com/94965944/194460354-280f9c51-99e4-411e-8591-6247ac439b80.png)


서비스 사용자가 10000000명으로 늘어날 때 쿠폰의 개수는 1억개가 등록되고 (인당 받은 쿠폰 약 10개)

쿠폰당 예약은 3억개, 쿠폰당 만남은 1억개

조직 수 : 약 2000개

조직의 회원 수 : 5000명

회원 조회 이름순으로 빠르게 조회해야한다.

회원을 빠르게 검색할 수 있어야한다.

- 검색 엔진

쿠폰 / 예약 / 만남 관련 Slack 알람이 1초당 몇번?

- 메시지큐 : async / non-blocking

스케줄러

- 배치

## 결론
## 너무 어렵다 ㅠㅠㅠㅠㅠㅠㅠㅠㅠㅠㅠㅠㅠㅠㅠㅠㅠㅠ

**Comment by tonic523** [comment]
## 스모디
![image](https://user-images.githubusercontent.com/59171113/194460036-14f42f6e-09ec-47fc-9150-d2948ed72180.png)

![image](https://user-images.githubusercontent.com/59171113/194460954-5348935e-6f3f-4954-a3ec-dcec34e18270.png)

**Comment by RIANAEH** [comment]
# 모락

<img width="1230" alt="image" src="https://user-images.githubusercontent.com/45311765/194460918-0c4b8f4b-b6b2-43d2-9e2e-b9f38017fc9d.png">

> 사용자 많아지면 어디가 문제될 것 같아요?

- 마감한 투표/약속잡기는 writer DB에서 관리하지 않는다.
    - 데이터가 많을 수록 INSERT, DELETE 시 인덱스 조정 비용이 높다.
    - 필요 없는 데이터는 writer DB에서 삭제하게 해 비용을 줄인다.
    - 배치를 이용해 하루에 한번씩 새벽 4시 쯔음에 불필요한 데이터를 writer DB에서 제거한다.
- 투표/약속잡기 생성, 마감 시 바로 사용자에게 알림을 보내야한다.
    - 메시지 큐를 이용해 알림을 보낸다.
- 약속잡기 추천 결과를 캐시에 저장한다.
    - 데이터를 빨리 보여줄 수 있다.
    - 서버의 불필요한 부하를 줄일 수 있다. (O)
    - 메시지 큐를 이용해 약속잡기 선택 이벤트를 받아 캐시 데이터를 갱신한다.
    - 캐시 쓰기 비용이 크다면 효율이 나쁠 수 있다.
    - 캐시 서버에서 동시성 문제가 발생할 수 있다.

**Comment by yaho99** [comment]
## 티타임

### 상황 가정

- 서비스 사용자가 1000만명으로 늘어나면 하나의 예약마다 5개의 슬랙 알람 메시지 즉, 5000만개의 슬랙 알람 메시지가 전송된다.
- 피크 시간에 예약이 몰리는 경우가 발생한다. 리마인더 알람이 가는 시간에는 약 만개의 예약에 해당하는 2만개의 알람 메시지가 일괄 전송된다.

---

- 개선한 인프라 아키텍처 이미지
  ![IMG_7225](https://user-images.githubusercontent.com/69156709/194460920-b070ee83-f4f8-42fe-b8b7-24915e9fef99.jpg)

**Comment by cndqjacndqja** [comment]
# 공책
![image](https://user-images.githubusercontent.com/48307960/194460976-8e68d9a6-64c6-4e5a-ada1-58ca94901db4.png)

## 예시

* 동시 사용자 10만 명이라고 가정할 경우, 10만 개의 커넥션을 서버에서 유지할 수 있어야한다.(Websocket)
* 서비스 사용자가  1000만 명이라고 가정할 경우, 완료된 제출이 하루 100만 건식 저장된다.
* 작업 별 완료된 제출을 시간 순으로 페이징해서 보여줘야한다.

실시간 서버를 분리하여 실시간으로 처리해야 할 부분과 HTTP 통신하는 부분을 구분한다.
HTTP 통신만 로드 밸런서를 적용.
job PK 해시값으로 RunningTask 파티셔닝.

**Comment by dongho108** [comment]
## 터놓고

### BE 열음 바니 수달 애쉬 앤지

### 서비스 데이터의 특성 (위의 예시와 같이 가정한 내용)

터놓고 타겟 대상이 우테코에서 → 교육기업으로 확대됨. 

- 각 교육기업에서 면담서비스를 이용하게 되었다.
- 어쩌다보니 회원 수 1000만명이 되었다.
- 어쩌다보니 회원 수 1000만명 서비스 기업의 사장?!
- 사용자가 면담을 엄청 많이해서 면담 데이터의 개수가 늘어났다.

### 우리 서비스에서 특별히 고려해야 할 중요한 기능

- 지난 면담 내역 데이터는 변경이 일어나지 않는다.
- 메일 발송, 슬랙 알림 발송을한다.
- 코치정보의 경우 변경이 잘 일어나지 않는다.
- 만일 다양한 교육기업이 도입 된다면, 기업별로 데이터 관리 포인트를 나눌 수 있다.

### 개선안 & 문제 & 해결방안

1. **캐싱 서버를 둔다.**
    - 요청 단위 캐시 vs 데이터 변경 빈도에 따른 캐시
        - 요청 단위 캐시 ((Write-Throgh) 방식을 사용한다.
            - 캐시서버에 요청 오는거 기준으로 응답값 캐싱 하고
            - 만약 그 데이터가 업데이트 되면 DB 업데이트 하면서 Cache 내의 데이터도 갱신
        - 데이터 변경 빈도에 따른 캐시
            - 지난 면담 내역과 같이 바뀌지 않는 데이터를 캐시한다.
    - Redis
        - In-Memory DB

1. **메세지 큐를 사용한다.**
    - 카프카 or rabbitMQ
        - pub, sub
            - publisher(WAS)에서 알람 이벤트 발행
            - subscriber(메세지 서버) 가 하나씩 가져감
        - 비동기여서 성능상의 이점이 있다.
        - 여유로울 때 가져간다.
        - 알림서비스 개발할 때
    - 파이어베이스
        - 경량 솔루션
    - 슬랙 알람 보낸다.
        - 알람은 갔는데, 면담이 생성이 안된다.
        - 면담은 생성 됐는데, 알람이 안간다.
        - 면담 생성 됐을 때, 알람이
        - 예외가 발생하면 알람 이벤트 생성을 하지 않는다.

1. **DBMS - 샤딩과 파티셔닝을 고려한다.** 
    - 파티셔닝을 먼저 도입한다.
        - 하드웨어가 파티셔닝으로 감당이 불가능한 경우 샤딩을 도입한다.
    - 회사별로 샤딩을 도입해 DBMS를 분리한다.
        - 관리비용이 늘어난다. (동기화 문제, join 성능 문제)
        - 데이터 조립이 힘들다.
        - 다른 테이블간의 의존이 없고
        - 부하가 심할 때 DBMS를 나눠서 쓰기 관련 부하를 나누고 싶을 때 샤딩을 쓰는건가?
        - 한 DB가 처리할 수 없을만큼 데이터가 많아질 때 어쩔 수 없을 때 쓴다.
        - 서비스가 10년째 이용이 됐을 때 연도별로 나눈다.
            - 리디북스 예시
                - 책이 200권 있는데 한권 찾기 힘들다.
                - 단행본(카테고리 별, 연도 별) 으로 나눠줄 수있나?

**Comment by OzRagwort** [comment]
# 레벨로그
- 개선한 인프라 아키텍처 이미지

![image](https://user-images.githubusercontent.com/32123302/194462646-999a505c-6610-4b1e-aed4-6e6bd7156e52.png)

- 인프라 아키텍처를 변경하면서 고민한 부분
    - 서비스 데이터의 특성
        - 팀 생성 부터 인터뷰 종료까지 꾸준하게 데이터가 쓰이고 조회됨
        - 인터뷰 진행 중 쓰기 작업이 많이 발생함
        - 인터뷰가 종료되면 해당 팀에서는 쓰기 작업이 없다
    - 우리 서비스에서 특별히 고려해야 할 중요한 기능
        - 인터뷰 질문이 검색되어야 한다.
        - 인터뷰 질문이 피크 시간에 매우 많이 작성이 된다.
        - e.g., 문자열 검색이 되어야 한다, 사용자에게 실시간으로 알람을 보내줘야 한다, 사용자가 동시에 데이터를 변경할 수 없도록 막아야 한다, 메인 페이지에서 특정 데이터를 자주 조회한다 등
    - 개선한 인프라 아키텍처에서 발생할 수 있을 것으로 예상되는 문제
        - 검색이 느리다
    - 문제에 대한 해결 방안 아이디어
        - *검색을 ES를 통해 성능 향상을 할 수 있을 것 같다.*

**Comment by gudonghee2000** [comment]
# 달록
- 개선한 인프라 아키텍처 이미지
<img width="1180" alt="1" src="https://user-images.githubusercontent.com/71062817/194468219-6bea95c2-d043-4ec9-aea0-6f377a78ba08.png">

- 인프라 아키텍처를 변경하면서 고민한 부분
어떻게 하면 월별 일정 데이터를 빠르게 클라이언트에게 전달할수 있을지

- 서비스 데이터의 특성 (위의 예시와 같이 가정한 내용)
각 사용자별로 독립적인 월별일정을 가진다.

- 우리 서비스에서 특별히 고려해야 할 중요한 기능
외부 캘린더(구글 캘린더)의 일정 데이터 사용자에게 전달해야한다.
저장된 일정 데이터를 빠르게 조회 할 수 있어야한다.

- 개선한 인프라 아키텍처에서 발생할 수 있을 것으로 예상되는 문제
외부 캘린더 데이터를 가져오면서 데이터가 누락될수 있다.
캐싱으로인한 관리점이 하나 더 늘어난다.

- 문제에 대한 해결 방안 아이디어
사용자의 외부 캘린더 데이터를 캐싱해놓는다.
데이터 베이스 파티션을 통해 조회 성능을 향상시킨다.
